src/main/java/com/avispl/symphony/dal/communicator/aver/ptz/AverPTZCommunicator.java

```diff
@@ -0,0 +1,1885 @@
+						presetControl(PresetControl.RECALL, Integer.parseInt(value));
+					}
+					break;
+				}
+				default: {
+					throw new IllegalStateException("Unexpected value: " + command);
+				}
+			}
+		} catch (CommandFailureException ex) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("error: control command not found");
+			}
+			throw ex;
+		} finally {
+			controlOperationsLock.unlock();
+		}
+	}
+
+	/**
+	 * This method is recalled by Symphony to control a list of properties
+	 *
+	 * @param controllableProperties This is the list of properties to be controlled
+	 */
+	@Override
+	public void controlProperties(List<ControllableProperty> controllableProperties) {
+		if (CollectionUtils.isEmpty(controllableProperties)) {
+			throw new IllegalArgumentException("AverCommunicator: Controllable properties cannot be null or empty");
+		}
+
+		controllableProperties.forEach(this::controlProperty);
+	}
+
+	/**
+	 * This method is recalled by Symphony to get the list of statistics to be displayed
+	 *
+	 * @return List<Statistics> This return the list of statistics.
+	 */
+	@Override
+	public List<Statistics> getMultipleStatistics() {
+		final ExtendedStatistics extStats = new ExtendedStatistics();
+		final Map<String, String> stats = new HashMap<>();
+		final List<AdvancedControllableProperty> advancedControllableProperties = new ArrayList<>();
+
+		if (this.cameraID < 1 || this.cameraID > 7) {
+			throw new IllegalArgumentException("Camera ID with value " + this.cameraID + " is out of range. Camera ID must between 1 and 7");
+		}
+
+		if (this.panSpeed < 0 || this.panSpeed > 24) {
+			throw new IllegalArgumentException("Pan speed with value" + this.panSpeed + " is out of range. Pan speed must between 0 and 24");
+		}
+
+		if (this.tiltSpeed < 0 || this.tiltSpeed > 24) {
+			throw new IllegalArgumentException("Tilt speed with value" + this.tiltSpeed + " is out of range. Tilt speed must between 0 and 24");
+		}
+
+		if (this.zoomSpeed < 0 || this.zoomSpeed > 7) {
+			throw new IllegalArgumentException("Pan speed with value" + this.zoomSpeed + " is out of range. Zoom speed must between 0 and 7");
+		}
+
+		if (this.focusSpeed < 0 || this.focusSpeed > 7) {
+			throw new IllegalArgumentException("Tilt speed with value" + this.focusSpeed + " is out of range. Focus speed must between 0 and 7");
+		}
+
+		// Monitoring capabilities
+		try {
+			initAverRestCommunicator();
+			deviceInfo = this.restCommunicator.getDeviceInfo();
+			populateMonitorCapabilities(stats);
+			// Control capabilities
+			populateControlCapabilities(stats, advancedControllableProperties);
+		} catch (Exception e) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("error: Cannot get data from Rest communicator: " + this.host + " port: " + this.port);
+			}
+			throw new ResourceNotReachableException("Aver rest communicator not reachable for getting data", e);
+		} finally {
+			try {
+				restCommunicator.disconnect();
```

Comment:
honestly, we do not need to call disconnect, but it is necessary to set login/password every time  (in case the user changes the login/password).
override authenticate() -> my adapter extends UDP communicator and can not do this, what do you mean? 😄

Reply:
I will create an AverPTZRestCommunicator instance in my main adapter's constructor and use it along with my main adapter (not call disconnect anymore).
when the user changes the login/password, the main adapter inits again -> create a new instance of AverPTZResetCommunicator.
do you think it is ok?